In [1]:
from __future__ import absolute_import, division, print_function, unicode_literals

import tensorflow as tf
from tensorflow import keras

import numpy as np

print(tf.__version__)

1.12.0


In [2]:
%run read_part_class.ipynb

In [3]:
print(gen_status_ee[1])
print(gen_id_ee[1])

print(gen_status_ee[0][5])

# for i in range(0,len(gen_status)):
#     for j in range(0,len(gen_status[i])):
#         if j != 91:
#             np.delete(gen_status[i], j)
# for i in range(0,4):
#     for j in range(0,len(gen_status[i])):
#         if j != 91:
#             print(np.delete(gen_status[i], j))
# x = np.arange(10)
# condlist = [x<3, x>5]
# choicelist = [x, x**2]
# np.select(condlist, choicelist)
# for i in range(0,4):
#     for j in range(0,len(gen_status[i])):
#         condlist = [gen_status[i][j] == 91]
#         choicelist = [gen_status[i][j]]
#         print(np.select(condlist, choicelist))
print(gen_status_ee[1])

[-22 -44 -44 -44 -62 -62 -62 -62 -62 -62 -62 -62 -62 -62 -62 -62 -62 -62
 -62 -63 -63 -63 -63  51  51  51  52  83  83  83  83  83  84  84  84  84
  84  84  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91
  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91
  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91
  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91  91
  91  91  91  91  91  91  91  91]
[   23    23    23    23    23    21    21    21    21    21    21    21
    21    21    21    21    21    21    21  2103  2101     2     2    22
   -11    22    11   321  2112 -2212  -211   211  -321   211   321   211
  -211   321  2212   211  -211   211  -211   211   321  -211   130  -321
   211   211  -211  -211  -211   211  -211   211   130   130    22    22
    22    22   211  -211    22    22    22    22    22    22    22    22
    22    22    22    22    22    22  2112  -211   211  -211    22    22
    22    22    2

In [4]:
# gen_id_ee = gen_id_ee + 5000
# gen_id_bb = gen_id_bb + 5000

print(gen_id_ee)
print(gen_id_bb)

[[23 23 23 ... 22 22 22] [23 23 23 ... 22 22 22] [23 23 23 ... -211 211 -211] ... [23 23 23 ... 22 22 22] [23 23 23 ... 211 211 -211] [23 23 23 ... 22 22 22]]
[[23 23 23 ... -211 211 -211] [23 23 23 ... 22 2212 22] [23 23 23 ... -211 22 22] ... [23 23 23 ... 22 22 22] [23 23 23 ... 22 -211 130] [23 23 23 ... 211 22 22]]


In [5]:
# print(gen_status_bb[1])
# print(gen_id_bb[1])

# print(gen_status_bb[0][5])

# print(gen_status_bb[1])

print(len(gen_id_bb))
sizes_bb = []
for i in range(0,len(gen_id_bb)):
    sizes_bb.append(len(gen_id_bb[i]))
print(np.amax(sizes_bb))

7130
1580


In [6]:
# # arr = np.array([[1,2,3,4], [5,6,7,8], [9,10,11,12]])
# # print(arr)
# # print(np.delete(arr, 1, 0))

# a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])
# # index = [2]

# # new_a = np.delete(a, index)

# # print(new_a.pop(4)) #Prints `[1, 2, 5, 6, 8, 9]
# import numpy.ma as ma
# # mx = ma.masked_array(a, mask = ((a < 3 ) | (a > 7)) )
# # print(mx)

# for i in range(0,4):
#     mx = ma.masked_array(gen_staus[i], mask = ( gen_status[i] == 91 )

# print(mx)
# ------------------------------------------------------------------------------------
# for i in range(0,4):
#     for j in range(0,len(gen_status_ee[i])):
#         if gen_status_ee[i][j] == 91:
# #             gen_status[i][j] = np.array(gen_status[i][j])
#             print(gen_status_ee[i][j])
# print(gen_status_ee)
# ------------------------------------------------------------------------------------
# gen_status_final = np.array
# print(gen_status_final)
# for i in range(0,4):
#     for j in range(0,len(gen_status[i])):
#         if gen_status[i][j] == 91:
#             gen_status_final[i][j] = np.array(gen_status[i][j])
#             print(gen_status_final[i][j])
# print(gen_status_final)


print(gen_status_ee[1][:] == 91)

[False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True]


In [7]:
is_final = abs(gen_status_ee) == 91
final_particles = particles[is_final]
# for event in range(0,len(gen_id)):
#     for part in range(0,len(gen_id[event])):
#         if gen_status.any == 91:
#             print(gen_id[part])

print(len(final_particles[0]))

346


**Still need to figure out how to select final state (91) only**

In [8]:
# #id_bins = np.linspace(0,2000,2001)
# # id_bins = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,20,21,22,23,24,25,30,100,220,240,260,300,340,380,420,480,500]
# id_bins = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,20,21,22,23,24,25,30]
# plt.hist(x=gen_id_ee.flatten(), bins=id_bins, log=True, density=True)

# # print(pairings_added.mass.flatten())

# plt.show()

In [9]:
# plt.clf()

# plt.hist(x=gen_id_bb.flatten(), bins=id_bins, log=True, density=True)

# plt.show()

In [10]:
class_names = ['Z_ee_jet', 'Z_bb_jet']
print(class_names)
Z_ee_jet = np.array([0]*len(gen_id_ee))
Z_bb_jet = np.array([1]*len(gen_id_bb))

print(Z_ee_jet)
print(len(Z_ee_jet))
print(Z_ee_jet.shape)
print(len(gen_id_ee))
print(gen_id_ee.shape)

print(Z_bb_jet)
print(len(Z_bb_jet))
print(Z_bb_jet.shape)
print(len(gen_id_bb))
print(gen_id_bb.shape)

['Z_ee_jet', 'Z_bb_jet']
[0 0 0 ... 0 0 0]
2574
(2574,)
2574
(2574,)
[1 1 1 ... 1 1 1]
7130
(7130,)
7130
(7130,)


In [11]:
# model = tf.keras.Sequential()

# model.add(keras.layers.Embedding(10000, 16))
# model.add(keras.layers.GlobalAveragePooling1D())
# model.add(keras.layers.Dense(16, activation=tf.nn.relu))
# model.add(keras.layers.Dense(1, activation=tf.nn.softmax))

# model.summary()

In [12]:
# model.compile(optimizer='adam',
#               loss='binary_crossentropy',
#               metrics=['acc'])

In [13]:
# model.predict(gen_id, Z_ee_jet, epochs=5)

In [14]:
part_size = 2048

# model = keras.Sequential()
# model.add(keras.layers.Embedding(part_size+1, 16))
# model.add(keras.layers.GlobalAveragePooling1D())
# # model.add(keras.layers.Dense(16, activation=tf.nn.relu))
# model.add(keras.layers.LeakyReLU(alpha=0.3))
# model.add(keras.layers.Dense(1, activation=tf.nn.sigmoid))
# # model.add(keras.layers.Dense(1, activation=tf.nn.softmax))

model = keras.Sequential([
#     keras.layers.Flatten(input_shape=(28, 28)),
#     keras.layers.Dense(128, activation=tf.nn.relu),
#     keras.layers.Dense(128, activation=tf.nn.tanh),
    keras.layers.Dense(256, activation=tf.nn.tanh),
#     keras.layers.LeakyReLU(alpha=0.3),
    keras.layers.Dense(16, activation=tf.nn.softmax)
])

# model.summary()

In [15]:
# model.compile(optimizer='adam',
#               loss='binary_crossentropy',
#               metrics=['acc'])
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['acc'])

changed compile params due to: https://stackoverflow.com/questions/50768538/indices-not-in-range-running-lstm-model-on-sequence-data

binary led to ${\approx}$30% accuracy, sparse gives ${\approx}$70% currently, need to play around with

In [16]:
# #print(gen_id[0])
# print(len(gen_id))
# print(len(gen_id)/2)

# half = int(len(gen_id)/2)

# x_val = gen_id[:half]
# partial_x_train = gen_id[half:]

# y_val = gen_id[:half]
# partial_y_train = gen_id[half:]

# print(x_val)
# print(partial_x_train)
# print(len(x_val))
# print(len(partial_x_train))
# print(len(gen_id[0]))

In [17]:
print(np.amax(len(gen_id_ee)))
print(len(gen_id_ee))
sizes_ee = []
for i in range(0,len(gen_id_ee)):
    sizes_ee.append(len(gen_id_ee[i]))
print(np.amax(sizes_ee))

print(len(gen_id_bb))
sizes_bb = []
for i in range(0,len(gen_id_bb)):
    sizes_bb.append(len(gen_id_bb[i]))
print(np.amax(sizes_bb))

2574
2574
1318
7130
1580


**Largest size is 1318 elements...so padd up 10 1320 ?**

**For now...haven't cut down to include only final stat yet**

**Largest size is now 1580**

In [18]:
max_len = -1

if (np.amax(sizes_ee)) >= (np.amax(sizes_bb)):
    max_len = np.amax(sizes_ee)
else:
    max_len = np.amax(sizes_bb)

print(max_len)

gen_id_ee = keras.preprocessing.sequence.pad_sequences(gen_id_ee,
                                                       value=0,
                                                       padding='post',
                                                       maxlen=max_len + 1)

gen_id_bb = keras.preprocessing.sequence.pad_sequences(gen_id_bb,
                                                       value=0,
                                                       padding='post',
                                                       maxlen=max_len + 1)

1580


In [19]:
# #print(gen_id[0])
# print(len(gen_id))
# print(len(gen_id)/2)

# half = int(len(gen_id)/2)

# # data
# x_val = gen_id[:half]
# partial_x_train = gen_id[half:]

# # labels of data
# y_val = Z_ee_jet[:half]
# partial_y_train = Z_ee_jet[half:]

# print(x_val)
# print(partial_x_train)
# print(len(x_val))
# print(len(partial_x_train))
# print(len(gen_id[0]))
# print(len(x_val[0]))
# print(len(partial_x_train[0]))
# print(x_val.shape)
# print(partial_x_train.shape)
# print(y_val)
# print(partial_y_train)
# print(y_val.shape)
# print(partial_y_train.shape)

#print(gen_id[0])
print(len(gen_id_ee))
print(len(gen_id_ee)/2)

print(len(gen_id_bb))
print(len(gen_id_bb)/2)

half_ee = int(len(gen_id_ee)/2)
half_bb = int(len(gen_id_bb)/2)

# data
x_val_ee = gen_id_ee[:half_ee]
partial_x_train_ee = gen_id_ee[half_ee:]

x_val_bb = gen_id_bb[:half_bb]
partial_x_train_bb = gen_id_bb[half_bb:]

# vstack ?

# labels of data
y_val_ee = Z_ee_jet[:half_ee]
partial_y_train_ee = Z_ee_jet[half_ee:]

y_val_bb = Z_bb_jet[:half_bb]
partial_y_train_bb = Z_bb_jet[half_bb:]

# vstack ?

# print(x_val)
# print(partial_x_train)
# print(len(x_val))
# print(len(partial_x_train))
# print(len(gen_id[0]))
# print(len(x_val[0]))
# print(len(partial_x_train[0]))
# print(x_val.shape)
# print(partial_x_train.shape)
# print(y_val)
# print(partial_y_train)
# print(y_val.shape)
# print(partial_y_train.shape)

2574
1287.0
7130
3565.0


In [20]:
print(x_val_ee)
print(partial_x_train_ee)
print(len(x_val_ee))
print(len(partial_x_train_ee))
print(len(gen_id_ee[0]))
print(len(x_val_ee[0]))
print(len(partial_x_train_ee[0]))
print(x_val_ee.shape)
print(partial_x_train_ee.shape)
print(y_val_ee)
print(partial_y_train_ee)
print(y_val_ee.shape)
print(partial_y_train_ee.shape)

[[23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 ...
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]]
[[23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 ...
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]]
1287
1287
1581
1581
1581
(1287, 1581)
(1287, 1581)
[0 0 0 ... 0 0 0]
[0 0 0 ... 0 0 0]
(1287,)
(1287,)


In [21]:
print(x_val_bb)
print(partial_x_train_bb)
print(len(x_val_bb))
print(len(partial_x_train_bb))
print(len(gen_id_bb[0]))
print(len(x_val_bb[0]))
print(len(partial_x_train_bb[0]))
print(x_val_bb.shape)
print(partial_x_train_bb.shape)
print(y_val_bb)
print(partial_y_train_bb)
print(y_val_bb.shape)
print(partial_y_train_bb.shape)

[[23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 ...
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]]
[[23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 ...
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]
 [23 23 23 ...  0  0  0]]
3565
3565
1581
1581
1581
(3565, 1581)
(3565, 1581)
[1 1 1 ... 1 1 1]
[1 1 1 ... 1 1 1]
(3565,)
(3565,)


In [22]:
x_val = np.vstack((x_val_ee, x_val_bb))
print(len(y_val_ee))
print(len(y_val_bb))
print(y_val_ee)
print(y_val_bb)
y_val = np.concatenate((y_val_ee, y_val_bb), axis=None)
print(y_val)
# y_val = np.vstack((y_val_ee, y_val_bb))

partial_x_train = np.vstack((partial_x_train_ee, partial_x_train_bb))
# partial_y_train = np.vstack((partial_y_train_ee, partial_y_train_bb))
partial_y_train = np.concatenate((partial_y_train_ee, partial_y_train_bb))

print(partial_y_train)

print(len(partial_y_train))

1287
3565
[0 0 0 ... 0 0 0]
[1 1 1 ... 1 1 1]
[0 0 0 ... 1 1 1]
[0 0 0 ... 1 1 1]
4852


Normalize the data ? Help with range issue?

In [23]:
# x_val = x_val / 10000
# partial_x_train = partial_x_train / 10000
# y_val = y_val / 10
# partial_y_train = partial_y_train / 10

x_val = x_val 
partial_x_train = partial_x_train 
y_val = y_val 
partial_y_train = partial_y_train 

**(Still need to figure out maxlen) -- Done(-ish)**

~~Possible error meanign for below block, re: array vs seq~~: https://www.google.com/search?client=ubuntu&channel=fs&q=setting+an+array+element+with+a+sequence.&ie=utf-8&oe=utf-8 -- **Resolved [x]**: (for model.fit, x is data and y is labels) https://keras.io/models/model/

So it looks like the problem is the gen-id somehow needs to be in the range of number of nodes? Negative values clearly weren't, so added 5000 to all to make all positive, but now same error as "indices[112,324] = 7212 is not in [0, 7131]" which is the same sort of error as before...am I confusing number of nodes with choices for id? Leaky ReLU handle negatives?

**Maybe I skipped an important preprocessing step? Check text-classification?**

In [24]:
history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=128,
                    validation_data=(x_val, y_val),
                    verbose=1)

Train on 4852 samples, validate on 4852 samples
Epoch 1/20
4852/4852 [==============================] - 1s 149us/step - loss: 2.3019 - acc: 0.3403 - val_loss: 1.5023 - val_acc: 0.5666
Epoch 2/20
4852/4852 [==============================] - 0s 101us/step - loss: 1.1718 - acc: 0.6321 - val_loss: 1.0127 - val_acc: 0.6430
Epoch 3/20
4852/4852 [==============================] - 0s 100us/step - loss: 0.8255 - acc: 0.6772 - val_loss: 0.8497 - val_acc: 0.6412
Epoch 4/20
4852/4852 [==============================] - 0s 91us/step - loss: 0.6943 - acc: 0.7042 - val_loss: 0.7809 - val_acc: 0.6570
Epoch 5/20
4852/4852 [==============================] - 0s 92us/step - loss: 0.6309 - acc: 0.7146 - val_loss: 0.7403 - val_acc: 0.6698
Epoch 6/20
4852/4852 [==============================] - 0s 90us/step - loss: 0.5911 - acc: 0.7312 - val_loss: 0.7199 - val_acc: 0.6655
Epoch 7/20
4852/4852 [==============================] - 0s 99us/step - loss: 0.5679 - acc: 0.7409 - val_loss: 0.7059 - val_acc: 0.6791
Epoc

In [25]:
print(partial_x_train[120][45])
print(partial_x_train.shape + x_val.shape)
print(partial_y_train.shape + y_val.shape)
print(len(gen_id_ee))
print(len(gen_id_bb))

21
(4852, 1581, 4852, 1581)
(4852, 4852)
2574
7130


In [27]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                multiple                  404992    
_________________________________________________________________
dense_1 (Dense)              multiple                  4112      
Total params: 409,104
Trainable params: 409,104
Non-trainable params: 0
_________________________________________________________________
